### DISCUSSION TITLE
Proper block dependency structure

### DISCUSSION THREAD

**Discussion by nbenn** [BristolMyersSquibb/blockr.core#1] [Ideas]
In v1, the primary dependency structure is a linear graph where we pass (within a stack) the result from one block to the next. This causes problems when introducing a "stack-linking" feature such as the `result_field` as we no longer have linear graphs but DAGs.

@DivadNojnarg with each block represented by a module and (IIUC) the preferred way of communication between modules being arguments to the "server-closure", do you have any ideas how this could look like? A block would have to be able to dynamically modify the arguments with which this "server-closure" is being called, no?

**Comment by DivadNojnarg** [comment]
@nbenn

> A block would have to be able to dynamically modify the arguments with which this "server-closure" is being called, no

Yes... 

Below a quick recap of what we currently have in v1 with out nested structure:

We nest blocks within a stack: 

```mermaid
flowchart TD
  subgraph stack[Stack]
    block1[block 1 module] --> |output blk1| block2[block 2 module] --> |...| blockn[block n module]
  end
```

Code wise it gives this (it's standard [documented](https://shiny.posit.co/r/articles/improve/communicate-bet-modules/) way with Shiny):

```r
stack_server <- function(input, output, session) {
  block1_output <- generate_server(block1)
  # block1_output$data must be a reactive
  block2_output <- generate_server(block2, in_dat = block1_output$data)
}
```

Now, by linking stack we want this:

```mermaid
flowchart TD
  subgraph workspace[Workspace]
    subgraph stack1[Stack1]
     direction TB
      block1[block 1 module] --> |output blk1| block2[block 2 module] --> |...| blockn[block n module]
    end
    subgraph stack2[Stack2]
     subgraph block3[block 3 module]
       result_field[Result field module]
     end
    end
    result_field -.-> | block3 requests stack 1 output| stack1_res
    stack1 --> stack1_res[Stack 1 result] -.-> block3
  end
```

Code wise, we now move at the workspace level:

```r
workspace_server <- function(input, output, session {
  stack1_output <- stack_server(stack1)
  stack2_output <- stack_server(stack2)
}
```

As the link is created at the field level (result field), how to properly surface this information? The classic shiny approach would be:
- The result field returns the selected stack id.
- The corresponding block returns an extra output (linked_to) containing the stack id.
- At the stack level we update a reactiveValue that is used as input of the targeted block. This value should contain the result of the linked stack.
- When the result field is empty, we destroy the link in a similar way.
- Overall, a block would have 2 data entry points: the previous block and an optional link.

And then, we'd have to handle the join block which is a bit special.

I believe things might be less complex if we were not having a nested structure for modules like you suggest in another comment
